### PR TITLE
#564: Collection Pipeline pattern

### PR DESCRIPTION
**Pull request by nikhilbarar**
Added an implementation of Collection Pipeline Pattern.

Added a new module 'collection-pipeline'.

### CODE REVIEW

**Comment by npathai** [comment]
@nikhilbarar Really sorry man, didn't get an opportunity this week to go through the PR. I will definitely complete it this weekend.

**Comment by nikhilbarar** [comment]
No issues @npathai :)

**Comment by iluwatar** [comment]
@nikhilbarar the PR has conflicts that should be resolved

**Comment by npathai** [comment]
@nikhilbarar @iluwatar I have picked up the review. In process, will be done by tomorrow.

File: collection-pipeline/README.md [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+Use the Collection Pipeline pattern when
+
+* When you want to perform a sequence of operations where one operation's collected output is fed into the next
+* When you use a lot of statements in your code
+* When you use a lot of loops in your code
+
+## Credits
+
+* [Function composition and the Collection Pipeline pattern](https://www.ibm.com/developerworks/library/j-java8idioms2/index.html)
+* [Martin Fowler](https://martinfowler.com/articles/collection-pipeline/)
```
Comment (npathai) [inline collection-pipeline/README.md]: We can add link to Java8 Streams as well
https://docs.oracle.com/javase/8/docs/api/java/util/stream/package-summary.html

File: collection-pipeline/src/main/java/com/iluwatar/collectionpipeline/App.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+  /**
+   * Program entry point.
+   * 
+   * @param args
+   *            command line args
+   */
+  public static void main(String[] args) {
+
+    List<Car> cars = Iterating.createCars();
```
Comment (npathai) [inline collection-pipeline/src/main/java/com/iluwatar/collectionpipeline/App.java]: Can we have a better name for `Iterating` class? Not sounding right.

File: collection-pipeline/src/main/java/com/iluwatar/collectionpipeline/Car.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+  private String model;
+  private int year;
+
+  /**
+   * Constructor to create an instance of car.
+   * @param theMake the make of the car
+   * @param theModel the model of the car
+   * @param yearOfMake the year of built of the car
+   */
+  public Car(String theMake, String theModel, int yearOfMake) {
```
Comment (npathai) [inline collection-pipeline/src/main/java/com/iluwatar/collectionpipeline/Car.java]: `the` prefix is not really needed here.

File: collection-pipeline/src/main/java/com/iluwatar/collectionpipeline/FunctionalProgramming.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+   * Method to get models using for collection pipeline.
+   * 
+   * @param cars {@link List} of {@link Car} to be used for filtering
+   * @return {@link List} of {@link String} representing models built after year 2000
+   */
+  public static List<String> getModelsAfter2000UsingPipeline(List<Car> cars) {
+    return cars.stream().filter(car -> car.getYear() > 2000)
+      .sorted(Comparator.comparing(Car::getYear))
+      .map(Car::getModel).collect(Collectors.toList());
+  }
```
Comment (npathai) [inline collection-pipeline/src/main/java/com/iluwatar/collectionpipeline/FunctionalProgramming.java]: Suggestions:
Your implementation is great, I was thinking if we can extend it a bit further.

We can add some other examples to better compare both styles

- Add `category` in Car class, and we can show the use of `groupingBy` collector, showing cars grouped by SUV, Sedan and such
- Add `Person` class with list of cars that the person owns and use `flatMap` to filter, sort and group cars.

File: collection-pipeline/src/main/java/com/iluwatar/collectionpipeline/CarFactory.java [hunk 5]
```diff
@@ -28,19 +28,19 @@
...
    * @return {@link List} of {@link Car}
    */
   public static List<Car> createCars() {
-    return Arrays.asList(new Car("Jeep", "Wrangler", 2011), new Car("Jeep", "Comanche", 1990),
-        new Car("Dodge", "Avenger", 2010),
-        new Car("Buick", "Cascada", 2016),
-        new Car("Ford", "Focus", 2012),
-        new Car("Chevrolet", "Geo Metro", 1992));
+    return Arrays.asList(new Car("Jeep", "Wrangler", 2011, "Jeep"), new Car("Jeep", "Comanche", 1990, "Jeep"),
+        new Car("Dodge", "Avenger", 2010, "Sedan"),
```
Comment (npathai) [inline collection-pipeline/src/main/java/com/iluwatar/collectionpipeline/CarFactory.java]: @nikhilbarar The implementation seems to be going good. Just an input, we should be using Enum for car category constants.

**Comment by nikhilbarar** [comment]
Hi @npathai I'm done with the changes. Please review and let me know if any other changes are needed. Thanks

File: collection-pipeline/README.md [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+* Java8 Streams (https://docs.oracle.com/javase/8/docs/api/java/util/stream/package-summary.html)
```
Comment (IAmPramod) [inline collection-pipeline/README.md]: Probably a typo, enclose `Java8 Streams` in square brackets.

**Comment by nikhilbarar** [comment]
@IAmPramod Done

File: collection-pipeline/src/test/java/com/iluwatar/collectionpipeline/AppTest.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+ * Tests that Collection Pipeline methods work as expected.
+ */
+public class AppTest {
+  
+  private List<Car> cars = CarFactory.createCars();
+  
+  @Test
+  public void testGetModelsAfter2000UsingFor() {
+    List<String> models = ImperativeProgramming.getModelsAfter2000(cars);
+    assertEquals(models, Arrays.asList("Avenger", "Wrangler", "Focus", "Cascada"));
```
Comment (npathai) [inline collection-pipeline/src/test/java/com/iluwatar/collectionpipeline/AppTest.java]: The arguments are in incorrect order. `assertEquals(expected, actual)`

File: collection-pipeline/src/test/java/com/iluwatar/collectionpipeline/AppTest.java [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+  }
+  
+  @Test
+  public void testGetModelsAfter2000UsingPipeline() {
+    List<String> models = FunctionalProgramming.getModelsAfter2000(cars);
+    assertEquals(models, Arrays.asList("Avenger", "Wrangler", "Focus", "Cascada"));
```
Comment (npathai) [inline collection-pipeline/src/test/java/com/iluwatar/collectionpipeline/AppTest.java]: Same as above, argument ordering issue in `assertEquals()`

File: collection-pipeline/src/test/java/com/iluwatar/collectionpipeline/AppTest.java [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+  }
+  
+  @Test
+  public void testGetGroupingOfCarsByCategory() {
+    Map<Category, List<Car>> modelsFunctional = FunctionalProgramming.getGroupingOfCarsByCategory(cars);
+    Map<Category, List<Car>> modelsImperative = ImperativeProgramming.getGroupingOfCarsByCategory(cars);
+    assertEquals(modelsFunctional, modelsImperative);
```
Comment (npathai) [inline collection-pipeline/src/test/java/com/iluwatar/collectionpipeline/AppTest.java]: Nit pick: Not an ideal way to test for output in general. If somehow defect gets replicated in both classes, the tests won't be able to catch that. A minor point not necessarily required in our case. I still if you feel it should be addressed. Let me know what you think.
It should be
```java
assertEquals(manualExpectedMap, modelsFunctional);
assertEquals(manualExpectedMap, modelsIterative);
```

File: collection-pipeline/src/test/java/com/iluwatar/collectionpipeline/AppTest.java [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+  }
+  
+  @Test
+  public void testGetSedanCarsOwnedSortedByDate() {
+    Person john = new Person(cars);
+    List<Car> modelsFunctional = FunctionalProgramming.getSedanCarsOwnedSortedByDate(Arrays.asList(john));
+    List<Car> modelsImperative = ImperativeProgramming.getSedanCarsOwnedSortedByDate(Arrays.asList(john));
+    assertEquals(modelsFunctional, modelsImperative);
```
Comment (npathai) [inline collection-pipeline/src/test/java/com/iluwatar/collectionpipeline/AppTest.java]: Same as previous comment

**Comment by nikhilbarar** [comment]
@npathai Done. Thanks for the review 👍

**Comment by npathai** [comment]
@nikhilbarar PNG image is still outdated. Please regenerate it and also add associations wherever possible between classes. Make sure UCLS file is also updated.

- [ ] Update class diagram PNG file
- [ ] Make sure ucls is updated

**Comment by nikhilbarar** [comment]
@npathai Updated UCLS and PNG

**Pull request merged**